### PR TITLE
Add ability to receive login code over SMS

### PR DESCRIPTION
**Pull request by tmb**
This PR allows users with a valid phone number attached to their account to receive their login code using SMS.

It adds two environment variables:
- `TWILIO_ACCOUNT_SID`
- `TWILIO_AUTH_TOKEN`

Example request:
```
PATCH v1/users/[id]

{"phone_number":"+12025550191", "auth_type":"sms"}
```

`auth_type` can either be `sms` or `email`. It can't be set to `sms` without a valid `phone_number` added to the user.

### CODE REVIEW

**Comment by zanedb** [comment]
Should `FROM_NUMBER` be an environment variable?

File: lib/twilio_client.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    AUTH_TOKEN = Rails.application.secrets.twilio_auth_token
+    FROM_NUMBER = '+14158532126'
+
+    TWILIO = Twilio::REST::Client.new(ACCOUNT_SID, AUTH_TOKEN)
+
+    def send_login_code(user)
+      TWILIO.messages.create(
+          from: FROM_NUMBER,
+          to: user.e164_phone_number,
+          body: "Hi! Someone (hopefully you!) requested a login code for Hack Club.\nYour login code is #{user.pretty_login_code}. It will expire in 15 minutes."
```
Comment (lachlanjc) [inline lib/twilio_client.rb]: Request: put the login code at the beginning so I don't have to open Messages to see the code.

Comment (tmb) [reply lib/twilio_client.rb]: fixed!

**Comment by tmb** [comment]
> Should `FROM_NUMBER` be an environment variable?

ehhh probably

**Comment by MaxWofford** [comment]
What happens if someone types "unsubscribe" to the twillo number. Are they locked out of their account?

**Comment by tmb** [comment]
> What happens if someone types "unsubscribe" to the twillo number. Are they locked out of their account?

It'll fallback to email if Twilio returns any errors. Going to implement that later today.

File: lib/twilio_client.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+module TwilioClient
+  class << self
+    ACCOUNT_SID = Rails.application.secrets.twilio_account_sid
+    AUTH_TOKEN = Rails.application.secrets.twilio_auth_token
+    FROM_NUMBER = '+14158532126'
```
Comment (zachlatta) [inline lib/twilio_client.rb]: Don't hardcode values.

Comment (tmb) [reply lib/twilio_client.rb]: oops, meant to fix that. will do.

**Comment by tmb** [comment]
@zachlatta finally finished getting that fixed, are we good to merge?

**Comment by lachlanjc** [comment]
@zachlatta What's happening here?

File: app/controllers/v1/users_controller.rb [hunk 3]
```diff
@@ -21,7 +21,10 @@ def auth
...
 
       if user.save
-        ApplicantMailer.login_code(user).deliver_later
+        if user.sms? && TwilioClient.send_login_code(user)
+        else
+          ApplicantMailer.login_code(user).deliver_later
+        end
 
         render_success(
           id: user.id,
```
Comment (zachlatta) [inline app/controllers/v1/users_controller.rb]: Response should differentiate with what kind of login code was sent (SMS or email).

Comment (zachlatta) [reply app/controllers/v1/users_controller.rb]: So we can display on the frontend where the login code was sent to

File: app/models/user.rb [hunk 4]
```diff
@@ -57,6 +70,7 @@ def default_values
 
     self.email_on_new_challenges ||= false
     self.email_on_new_challenge_posts ||= false
+    self.auth_type = 'email'
```
Comment (zachlatta) [inline app/models/user.rb]: Should be `||=`

File: app/models/user.rb [hunk 5]
```diff
@@ -105,6 +119,10 @@ def pretty_login_code
     login_code&.scan(/.../)&.join('-')
   end
 
+  def e164_phone_number
+    sms? ? Phonelib.parse(phone_number).e164 : nil
```
Comment (zachlatta) [inline app/models/user.rb]: Why not just check to see if they have a phone number?

File: spec/requests/v1/users_spec.rb [hunk 6]
```diff
@@ -341,6 +341,54 @@
...
+              auth_type: 'sms'
+            }
+          end
+
+          it 'gracefully fails' do
+            expect(response.status).to eq(422)
+            expect(json['errors']).to include('auth_type')
+          end
+        end
+
```
Comment (zachlatta) [inline spec/requests/v1/users_spec.rb]: Should add tests for login code generation.

**Comment by lachlanjc** [comment]
Closing for now, since will require changes on Apply & Bank & this PR isn’t ready. Would still love to implement sometime!

**Pull request closed without merging**